### PR TITLE
Restrict KV endpoint to root and node users.

### PR DESCRIPTION
**Pull request by mberhault**
Work towards #2089
- restrict all RPC endpoints to root and node users only.
- restrict http kv endpoint to root and node users only.
- sql http endpoint allows all users

Nothing makes use of the rpc.RegisterPublic method yet, but sql will.

### CODE REVIEW

File: rpc/server.go [hunk 1]
```diff
@@ -94,9 +95,12 @@ func NewServer(addr net.Addr, context *Context) *Server {
 // argument of the same type as `reqPrototype`. Both the argument and
 // return value of 'handler' should be a pointer to a protocol message
 // type. The handler function will be executed in a new goroutine.
-func (s *Server) Register(name string, handler func(proto.Message) (proto.Message, error),
+// If 'restricted' is true, only system users (RootUser and NodeUser)
+// may call this method.
+func (s *Server) Register(name string, restricted bool,
```
Comment (petermattis) [inline rpc/server.go]: Rather than a `restricted` argument here, how about leaving `Register` as-is and adding a `RegisterPublic` method for use by `sql`.

**Comment by petermattis** [comment]
The structure looks fine to me, though I might use the term `public` instead of `restricted`.

File: rpc/server.go [hunk 2]
```diff
@@ -388,7 +396,10 @@ func (s *Server) readRequest(codec rpc.ServerCodec) (req rpc.Request, m method,
 	if ok {
 		args = reflect.New(m.reqType.Elem()).Interface().(proto.Message)
 	}
-	err = codec.ReadRequestBody(args)
+	if err = codec.ReadRequestBody(args); err != nil || args == nil {
```
Comment (tamird) [inline rpc/server.go]: nit: i'd put the `args != nil` check before the call to authHook rather than here

File: rpc/server.go (hunk 1 above)
Comment (tamird) [inline rpc/server.go]: +1

**Comment by tamird** [comment]
LGTM

File: rpc/server.go (hunk 2 above)
Comment (MBerhault) [inline rpc/server.go]: done

File: rpc/server.go (hunk 1 above)
Comment (MBerhault) [inline rpc/server.go]: done

**Comment by MBerhault** [comment]
adjusted for comments. Will ping again when this is ready.

**Comment by MBerhault** [comment]
This is ready. Except for addressing the review comments, all other changes are to tests.

**Comment by petermattis** [comment]
LGTM, though I didn't examine the test changes/coverage in depth.

File: server/authentication_test.go [hunk 3]
```diff
@@ -18,35 +18,67 @@
...
+			t.Fatal(err)
+		}
+		b = bytes.NewReader(buf)
+	}
+	req, err := http.NewRequest(method, url, b)
 	if err != nil {
 		t.Fatalf("%s %s: error building request: %s", method, url, err)
 	}
+	if b != nil {
+		req.Header.Add(util.ContentTypeHeader, util.ProtoContentType)
```
Comment (tamird) [inline server/authentication_test.go]: nit: any reason not to always do this?

**Comment by tamird** [comment]
LGTM

File: server/authentication_test.go (hunk 3 above)
Comment (MBerhault) [inline server/authentication_test.go]: It's accepted form that the content-type should be set when there is a body. When there is not, it doesn't actually hurt to set it, but there's really no reason to.

**Pull request merged**